jupyterhub/singleuser/mixins.py

```diff
@@ -707,6 +715,26 @@ def make_singleuser_app(App):
                 "{}.base_handler_class must be defined".format(App.__name__)
             )
 
+    # patch-in hub HubOAuthCallbackHandler to BaseHandler,
+    # so anything inheriting from BaseHandler uses Hub authentication
+    if HubAuthenticatedHandler not in BaseHandler.__bases__:
+        log.debug(f"Patching {HubAuthenticatedHandler} into {BaseHandler}")
+        BaseHandler.__bases__ = (HubAuthenticatedHandler,) + BaseHandler.__bases__
+        # adding it to bases isn't enough to override methods defined on BaseHandler, though.
+        # we still need to override any methods defined on BaseHandler *itself* that we should override
+        # since bases come immediately *after* the class itself
```

Comment:
#### Is this understanding correct?
- `the class itself` in the comment refers to BaseHandler
- The goal is not to redefine BaseHandler or similar, but to patch it if needed with adjustments defined in `HubAuthenticatedHandler`
- mro() (Method Resolution Order) is used on HubAuthenticatedHandler to select the correct method among the various potentially duplicate methods among the inheritance tree down to the BaseHandler class.
- Various attributes of BaseHandler won't be updated, only methods.

#### Question
- Do you feel confident that methods patched in BaseHandler with something from from HubAuthenticatedHandler or its base classes won't rely on some attribute not available in BaseHandler and would error out? Is this an actual constraint that merits a comment in HubAuthenticatedHandler and its base classes on top of BaseHandler?

Reply:
> #### Is this understanding correct?

> * `the class itself` in the comment refers to BaseHandler

Yes

> * The goal is not to redefine BaseHandler or similar, but to patch it if needed with adjustments defined in `HubAuthenticatedHandler`

Yes

> * mro() (Method Resolution Order) is used on HubAuthenticatedHandler to select the correct method among the various potentially duplicate methods among the inheritance tree down to the BaseHandler class.

This is how `self.method()` find out what `method` refers to - it looks for the first match in inheritance order.

> * Various attributes of BaseHandler won't be updated, only methods.

Not necessarily. Class attributes work the same way. Methods are callable attributes, so it is not useful to distinguish them.

> 
> #### Question

> * Do you feel confident that methods patched in BaseHandler with something from from HubAuthenticatedHandler or its base classes won't rely on some attribute not available in BaseHandler and would error out? Is this an actual constraint that merits a comment in HubAuthenticatedHandler and its base classes on top of BaseHandler?

HubAuthenticatedHandler inherits from HubAuthenticated, and only assumes that those attributes are defined. It relies on no methods or attributes from BaseHandler. It should work for any tornado RequestHandler.

The complication here is that HubAuthenticatedHandler is meant to define two methods in the public RequestHandler API:

- get_login_url
- get_current_user

Using these in standard inheritance, as should be done in any jupyterhub-authenticated service works without complication:

```python
class MyHandler(HubOAuthenticated, RequestHandler):
    hub_users = {'me'}
    @web.authenticated
    def get(self): ...
```

However, for singleuser we are patching-in the same logic to already-configured classes, so we need to apply these changes at runtime, hence insertion into `__bases__` and the extra fiddly bits. Ultimately, the goal is to behave *as if these were defined in BaseHandler*. I will try to simplify the logic here.